src/lib/io/sol-uart-impl-riot.c

```diff
@@ -117,50 +118,124 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
+    }
+
+    uart->tx_cb = config->tx_cb;
+    uart->user_data = config->user_data;
+    sol_ptr_vector_init(&uart->pending_blobs);
+    sol_buffer_init(&uart->rx_buffer);
+    uart->rx_buffer.flags |= SOL_BUFFER_FLAGS_NO_NUL_BYTE;
+
     return uart;
 
 fail:
+    uart_poweroff(uart->id);
     free(uart);
     return NULL;
 }
 
 SOL_API void
 sol_uart_close(struct sol_uart *uart)
 {
+    struct sol_blob *blob;
+    uint16_t i;
+
     SOL_NULL_CHECK(uart);
 
-    if (uart->async.tx_writer) {
-        sol_timeout_del(uart->async.tx_writer);
-        uart_tx_dispatch(uart, -1);
+    SOL_PTR_VECTOR_FOREACH_IDX (&uart->pending_blobs, blob, i) {
+        uart_tx_dispatch(uart, blob, -ECANCELED);
+        sol_blob_unref(blob);
     }
 
-    if (uart->async.handler)
-        sol_interrupt_scheduler_uart_stop(uart->id, uart->async.handler);
-    uart_poweroff(uart->id);
+    if (uart->tx_writer)
+        sol_timeout_del(uart->tx_writer);
+
+    if (uart->handler)
+        sol_interrupt_scheduler_uart_stop(uart->id, uart->handler);
 
+    uart_poweroff(uart->id);
+    sol_ptr_vector_clear(&uart->pending_blobs);
+    sol_buffer_fini(&uart->rx_buffer);
     free(uart);
 }
 
-SOL_API bool
-sol_uart_write(struct sol_uart *uart, const uint8_t *tx, unsigned int length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data)
+SOL_API int
+sol_uart_write(struct sol_uart *uart, struct sol_blob *blob)
 {
-    SOL_NULL_CHECK(uart, false);
-    SOL_EXP_CHECK(uart->async.tx_buffer, false);
+    bool created_writer = false;
+    int r;
 
-    uart->async.tx_buffer = tx;
-    uart->async.tx_cb = tx_cb;
-    uart->async.tx_user_data = data;
-    uart->async.tx_length = length;
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(blob, -EINVAL);
 
-    uart->async.tx_writer = sol_timeout_add(0, uart_tx_cb, uart);
+    if (!uart->tx_writer) {
+        uart->tx_writer = sol_timeout_add(0, uart_tx_cb, uart);
+        SOL_NULL_CHECK(uart->tx_writer, -ENOMEM);
+        created_writer = true;
+    }
 
-    return true;
+
+    r = -EOVERFLOW;
+    SOL_NULL_CHECK_GOTO(sol_blob_ref(blob), err_blob);
+    r = sol_ptr_vector_append(&uart->pending_blobs, blob);
+    SOL_INT_CHECK_GOTO(r, < 0, err_append);
+    return 0;
+
+err_append:
+    sol_blob_unref(blob);
+err_blob:
+    if (created_writer) {
+        sol_timeout_del(uart->tx_writer);
+        uart->tx_writer = NULL;
+    }
+    return r;
+}
+
+SOL_API int
+sol_uart_read(struct sol_uart *uart, struct sol_buffer *buf)
+{
+    size_t min;
+    int r;
+
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(buf, -EINVAL);
+
+    if (!uart->rx_buffer.used)
+        return 0;
+
+    if (buf->flags & SOL_BUFFER_FLAGS_FIXED_CAPACITY) {
+        min = sol_min(uart->rx_buffer.used, (buf->capacity - buf->used));
+        memcpy(sol_buffer_at_end(buf), uart->rx_buffer.data, min);
+    } else {
+        min = uart->rx_buffer.used;
+        r = sol_buffer_append_buffer(buf, &uart->rx_buffer);
```

Comment:
just a note, we'd have to guarantee all flags match before doing that, such as nul-byte.